### PR TITLE
Fix the fee calculation

### PR DESCRIPTION
**Pull request by glonlas**
This PR fixes the issues #176 and #141 

## Overview
Long story short, with this PR we change how the fee calculation is made.  
 
**Before:** The bot applied 2x0.25% on the closing rate, that is not how Bittrex applies fees. It means fee on the table column `close_profit`, are currently wrong.
**Now:**  By merging this PR, we apply the exact Bittrex fee calculation. 0.25% of the open rate then, 0.25% of the close rate. New closed trade will have the right `close_profit` column

## Quick changelog
* Change the Open and close trade mock to follow a real Bittrex transaction (See scenario below)
* Rename `Trade::calc_profit` to `Trade::calc_profit_percent`
* Refactor `Trade::calc_profit`
* Add fee calculation method breakdown: `Trade::calc_open_trade_price`, `Trade::calc_close_trade_price` and `Trade::calc_open_trade_price` that return the profit in BTC
* Add unitests
* Update existing unittest to ensure they expects a real transaction fees
* Update `/profit` and `/daily` to use `trade.calc_profit methods instead of re-implementing them

## Scenario
On this test, we will buy and sell a cryptocurrency.

**Buy**
* Buy: `90.99181073` Crypto at 0.00001099 BTC (`90.99181073 * 0.00001099` = 0.0009999 BTC)
* Buying fee: `0.25%`
* Total cost of buy trade: `0.001002500 BTC` (`(90.99181073 * 0.00001099) + ((90.99181073 * 0.00001099)* 0.0025)`)

**Sell**
* Sell: `90.99181073` Crypto at 0.00001173 BTC (`90.99181073 * 0.00001173` = 0,00106733394 BTC)
* Selling fee: `0.25%`
* Total cost of sell trade: `0.001064666 BTC` (`(90.99181073 * 0.00001173) - ((90.99181073 * 0.00001173) * 0.0025)`)

Profit/Loss: `+0.000062166` BTC (Sell:0.001064666 - Buy:0.001002500)
Profit/Loss percentage: `0.0620`  (`(0.001064666/0.001002500)-1` = `6.20%`)

## Warning
This PR does not include the DB update to make previous `close_profit`data right. A script to fix the DB will come later. Between we will have a mix of old incorrect data, and new correct data.

Because the bot based all his decision on the estimated profit, your strategy will be affected. Please re-run your backtesting to ensure you are still using the right value for your strategy.

## Backtesting `avg profit` and `total profit` will change
Because until now the fee calculation was wrong, the avg_profit was wrong on the backtesting report. With this PR we restore the right fee calculation, do not be surprised to see an update on your `avg profit` and `total profit`

This PR fixes the wrong amount of coin bought by the backtesting. (Explained on the issue #177)
At the same time, this PR is using the new profit calculation the PR introduces.

### Why it was wrong before?
Sample of the new result if the `stake_amount`= 0.001 BTC with `max_open_trades` =15, I cannot expect to earn 1 BTC or more within a month.

### How do we know we are good now?
An example that shows we are now on a good track: 
```
pair         buy count  avg profit    total profit       avg duration
---------  -----------  ------------  ---------------  --------------
BTC_ETH              1  1.79%         0.00001797 BTC            35
```
The bot has invested 0.001 for only 1 trade and made ~1.79% of profit, the result should be `0.00000179` (0.001*0.00179 = 0.00000179), not a huge result has the current backtesting is giving to us.

## New backtesting result

```
====================== BACKTESTING REPORT ======================================
pair         buy count  avg profit    total profit       avg duration
---------  -----------  ------------  ---------------  --------------
BTC_ADA              2  7.82%         0.00014719 BTC            10
BTC_NEO              5  1.97%         0.00009683 BTC            39
BTC_NXT              2  3.13%         0.00005874 BTC            15
BTC_MCO              9  1.13%         0.00008361 BTC            28.33
BTC_ETH              1  1.79%         0.00001797 BTC            35
BTC_BCC              3  1.81%         0.00005304 BTC            70
BTC_VOX              2  8.68%         0.00015672 BTC            10
BTC_GUP              3  -0.22%        -0.00001016 BTC           38.33
BTC_SC               0  nan%          nan BTC                  nan
BTC_VTC              5  1.80%         0.00008822 BTC            72
BTC_STRAT            4  4.71%         0.00018345 BTC            51.25
BTC_OMG              1  1.66%         0.00001598 BTC            35
BTC_OK               4  1.23%         0.00004768 BTC            62.5
BTC_EDG              7  3.11%         0.00021141 BTC            53.57
BTC_STORJ            2  2.35%         0.00004502 BTC            75
BTC_EMC2             0  nan%          nan BTC                  nan
BTC_XLM              1  2.37%         0.00002323 BTC            25
BTC_LSK              5  2.87%         0.00013736 BTC            32
BTC_SYS              5  2.68%         0.00012907 BTC            32
BTC_POWR             3  5.48%         0.00014186 BTC            16.67
BTC_PAY              8  -2.24%        -0.00017582 BTC          173.12
BTC_DGB              0  nan%          nan BTC                  nan
BTC_ETC              1  1.25%         0.00001219 BTC            50
BTC_XRP              3  -0.14%        -0.00000503 BTC           25
BTC_LTC              1  2.17%         0.00002136 BTC            35
BTC_IOP              2  2.86%         0.00005532 BTC            27.5
BTC_RCN              2  3.27%         0.00006376 BTC            30
BTC_BTG              2  1.15%         0.00002180 BTC            77.5
BTC_MONA             7  3.03%         0.00020639 BTC            51.43
BTC_SALT            13  0.50%         0.00006609 BTC            75
BTC_DASH             5  -1.43%        -0.00007039 BTC           78
BTC_QTUM             3  -1.23%        -0.00003647 BTC           75
BTC_CVC              4  2.47%         0.00009543 BTC            40
BTC_KMD              8  3.04%         0.00023851 BTC            67.5
BTC_XEM              4  -4.05%        -0.00015934 BTC           95
BTC_XMR              5  -0.63%        -0.00002984 BTC           59
BTC_ZEC              8  0.87%         0.00006885 BTC            52.5
BTC_WAVES           10  -0.57%        -0.00005667 BTC           76
BTC_PIVX             2  3.46%         0.00006616 BTC            22.5
BTC_XZC             13  1.30%         0.00016428 BTC            56.54
TOTAL              165  1.39%         0.00217380 BTC            59.09
```

### CODE REVIEW

**Comment by chrisapril** [comment]
:+1:

File: freqtrade/persistence.py [hunk 1]
```diff
@@ -119,13 +121,69 @@ def close(self, rate: float) -> None:
...
+        :return: Price in BTC of the open trade
+        """
+        getcontext().prec = 8
+
+        buy_trade = (Decimal(self.amount) * Decimal(self.open_rate))
+        fees = buy_trade * Decimal(fee or self.fee)
+        return float(buy_trade + fees)
+
+    def calc_close_trade_price(
+            self, rate: Optional[float] = None, fee: Optional[float] = None) -> float:
```
Comment (shusso) [inline freqtrade/persistence.py]: IIRC the pythonic way would be 
```
def calc_close_trade_price(self,
                                              rate ..,
                                              ..) -> float:
```

**Comment by glonlas** [comment]
@shusso Here we go, I made the change :)

**Comment by vertti** [comment]
This will mess up my hyperopting objective function. I'll follow up with a fix for that.

**Comment by glonlas** [comment]
do not hesitate to tell me what is wrong now. I will be glad to fix it

**Pull request merged**